### DISCUSSION TITLE
In Object.defineProperty `this.push()` works but `this.concat()` does not work

### DISCUSSION THREAD

**Discussion by ganeshkbhat** [nodejs/help#5041] [Q&A] [answered]
### Version

18.10.0

### Platform

Microsoft Windows NT 10.0.22621.0 x64

### Subsystem

V8 engine Object prototype

### What steps will reproduce the bug?

Just run the following

```
function clear(start, end) {
    this.splice((!!start) ? start : 0, (!!end) ? end : this.length);
}

function extend(iterable, index) {
    let a = [...this];
    // this.clear();
    this.concat(...[...a.splice(0, index), ...iterable, ...a]); // does not work
    // this.push(..); // works
}

Object.defineProperty(Array.prototype, 'clear', { value: clear, enumerable: true, });
Object.defineProperty(Array.prototype, 'extend', { value: extend, enumerable: true, });

let a = [1, 2, 3, 4, 5, 6, 7, 8, 9];
let b = [1, 2, 3, 4, 5, [1, 2, 3], 6, 7, [1, 2, 3, 4, [1, 2, 3], 5, 6], 8, [1, 2, [1, 2, [1, 2, 3], 3, 4, 5], 3, 4, 5], 9];

let e = [...a];
console.log("extend :", e.extend(b, 3));
console.log("result", e);
```

### How often does it reproduce? Is there a required condition?

Always. No, this is javascript default.

### What is the expected behavior?

`this.concat()` shoudl be able to modify.

### What do you see instead?

`this.push()` works but `this.concat()` does not work.

### Additional information

`this.push()` works but `this.concat()` does not work. this looks like an internal bug. is this supposed to be an RFC? https://stackoverflow.com/questions/75418335/unable-to-use-this-concatarray-inside-an-custom-array-prototype-extend-functio

**Comment by cjihrig** [comment, marked as answer]
The second sentence of the [`Array.prototype.concat()` page on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat):

> This method does not change the existing arrays, but instead returns a new array.

This is different from `Array.prototype.push()`, which modifies the array.

Side note: This is not specific to Node.js, and so Node's issue tracker is not the right place for something like this in the future.

**Reply by ganeshkbhat** [reply]
correct. i noticed this now.

```
function sorters(key = null, reverse = false) {
    if (!!key && typeof key === "function") { 
         let a = this.map(key); 
    } else {
         let a = this.concat([1,2,3,4,5]);
    }
    this.length = 0;
    this.push(...a);
}

function mappers(func) {
    if (!!func && typeof func === "function") { let a = this.map(func); }
    else { let a = this.map((i) => { return i * 2 }) }
    this.length = 0;
    this.push(...a)
}
Object.defineProperty(Array.prototype, 'sorters', { value: sorters, enumerable: true, });
Object.defineProperty(Array.prototype, 'mappers', { value: mappers, enumerable: true, });
let c = [1, 100000, 21, 30, 4];

console.log(c);
console.log(c.mappers());
console.log(c);
console.log(c.map((i) => { return i * 2 }))
console.log(c)
```

I notice now. Seems like both `[].map()` and `[].concat()` works by returning a new array both alike; unlike `[].forEach()`. my bad. probably i got confused since `this` is an immutable internal and gives access to the array and i generally use `[].concat()` rather than `this.concat()`.

I was mistaking `[].concat` to be an mutater rather than a new object creator. On the second point, I was checking if I should be raising an RFC for such? what is the general process?
